### DISCUSSION TITLE
Add Pull request Template chooser similar to Issue Template chooser

### DISCUSSION THREAD

**Discussion by Andre601** [community/community#4620] [Pull Requests]
## The issue
We all know that Pull requests are important.
What we also know is, that not every Pull request is the same and that sometimes one template isn't enough to work with.

The big issue here is that, unlike issues, Pull requests do not have a Template chooser, which the person submitting the Pull request could select from.
The only way they can select a different template is by applying the `template=<template>.md` query parameter to the URL to load it, which is extremely inconvenient.

## The idea
There should be a Template chooser for Pull requests. This would allow maintainers to offer Pull request templates for specific types such as updating documentation.

As an added bonus could they also have a way to define the default branch to use, because often is the `master`/`main` branch not the one someone should PR towards and blocking the branch from direct PRs is often only a semi-optimal solution.
Instead, could the selected template set the branch to target, so that the user submitting it has less to worry about.

**Comment by Andre601** [comment]
I've edited my original post and added an additional idea for why the template chooser would be a beneficial idea.

**Comment by jmhwang7** [comment]
A template chooser would also be really beneficial in monorepos where different components may have different considerations to think about when making a change (how would we deploy this? reliability?, etc.)

**Reply by davwheat** [reply]
This is a major use case for us at Flarum, where we have one single `framework` repository which contains JS packages, PHP packages as well as our core composer package and core extensions.

**Comment by conhanley** [comment]
It would also be nice if the template chosen for the PR could be determined by a chosen label (e.g. choose Documentation label upon PR creation -> populate description with Documentation PR Template). 

"Feed two birds with one seed"

**Comment by ShadowMitia** [comment]
It would be very beneficial in the case of the [Arcane Algorithm Archive](github.com/algorithm-archivists/algorithm-archive/), where we have two major ways to contribute: chapters and code. Chapters are complex PRs that require time and effort to manage, while code PRs can range from fixing bugs, improving readability to converting to new languages not yet present in the book.

Being able to have templates for either (and possibly for more cases) would be helpful in instructing how to proceed in each case, but would also be beneficial for people not accustomed to contributing to give them the right pointers.

There are a lot of useful possibilites with this features!

**Comment by maxfie1d** [comment]
+1

**Comment by kenji-miyake** [comment]
FYI, this is my workaround. It's somewhat better than manually changing the URL parameter.
https://github.com/kenji-miyake/test-pr-template

https://user-images.githubusercontent.com/31987104/184344710-2adee239-799f-4fdf-bfab-be76345bfac1.mp4

I remember I posted this to some thread on https://github.community/, but it seems not to be transferred. So, I re-posted it here. I really want the PR template chooser feature officially.

**Reply by uncenter** [reply]
~~I'm struggling to get this to work over at https://github.com/uncenter/pull-request-template-test... it works when I fork your repo and propose changes but not on my end. Does it need to be a forked repo and not my own? Seems to work for you though.~~ Nevermind! Still works, thank you for the workaround.

**Reply by VictorIreri** [reply]
It wasn't working for me until I realised that I had made a small edit to the PR description before clicking Preview and trying the link. This is because GitHub remembers changes you make so it was restoring the exact state of the PR description after the refresh.

**Comment by lkraav** [comment]
I don't usually do this, but 2023 bump.

[Suggestion to use template query parameters](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/using-query-parameters-to-create-a-pull-request) means developers have to start manually fiddling with browser location bar every day? My brain doesn't want to believe a product team has considered this a realistic solution (even if temporary), so I must be missing something about available UX options?

**Comment by rashadksh** [comment]
I told ChatGPT to create me a chrome extension that injects a dropdown into pull request page that has the templates.

You can use it from here -> https://github.com/userashad/github-template-picker-chrome-extension

**Comment by bmitc** [comment]
The ability to do this already exists in the backend, since [pull request templates can be specified via query parameters](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) and the [GitHub CLI through `gh pr create --template <template>` also supports this](https://cli.github.com/manual/gh_pr_create). It seems to me that all that is required is to update the create pull request buttons to be similar to the new issue buttons.

**Comment by uncenter** [comment]
It's pretty frustrating that this is still not available (851 days after OP made this!). Seems like a small change as @bmitc said considering this already exists for issues and choosing a template already works (inconveniently) with query parameters.

**Comment by abstractalgo** [comment]
Bumping this one as the amount of work involved should be minimal compared to the value given to the customers (us, the developers).

**Comment by moshe-foreflight** [comment]
Bump

**Reply by moshe-foreflight** [reply]
Sorry for the terse comment. I'm bumping this because I work in an environment where the template is massive and many of the sections don't apply. 

Also, there seems to be a bug where the line immediately following a markdown header that it's impossible to type. I have no idea what is causing the issue but it's quite frustrating to have to hit `<enter>` twice to continue typing. I can file a bug elsewhere if it's more appropriate.

Thank you!

**Comment by AlekseiDE** [comment]
Bump

**Comment by null-sleep** [comment]
Bump

**Comment by Andre601** [comment]
I would really apreciate if you guys stoped posting a stupid "bump" message. It won't get this worked on faster, nor gain the attention of GitHub Staff faster.

If that continues will I request a locking of this discussion by the repository admins. So take this as your only warning.

**Reply by bmitc** [reply]
I agree that the bump messages provide no benefit and should be stopped. At the same time, it's also disappointing that this suggestion has never once been addressed, much less considered, by the GitHub team in three years, especially given that this seems like it would be a very straightforward design, UX, and implementation while being extremely useful.

**Reply by conhanley** [reply]
I just found this Unsubscribe button at the bottom, and am using it. Apparently GitHub issues allow unsubscribing, just like PRs.

Being subscribed to emails from this issue won't get this worked on faster, nor gain the attention of GitHub Staff faster.

**Comment by conhanley** [comment]
What if ChatGPT just picked the template for you? I think I’d prefer a lack of choice now that technology and AI have advanced.

not a bump, as this is a genuine addition to the enhancement request. That said, I wouldn’t mind an admin taking a look at this post if it meant a little attention. “No press is bad press”, as they say…Unless I get banned, which would stink. Please don’t ban me…😅

**Comment by DangerBlack** [comment]
I don't think any of you need this but I've build the grease monkey equivalent of the things written by @rashadksh it's more stupid but get the job done. (you just configure it by typing your template names in an array)

https://github.com/DangerBlack/greasemonkey-scripts/tree/master/github-template-picker

**Comment by ppiont** [comment]
This functionality would be tremendously helpful for us.

**Comment by deitChi** [comment]
I've recently done some work on validating PR Template contents, which lead me to discover this post.

You can update the PR contents via [Pull Request API](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#update-a-pull-request) and updating the `body`.

So a simple work around might look like

```
#.gihub/pull_request_template.md
- [ ] PR for x
- [ ] PR for y
- [ ] PR for z
```

And an `update` action that replaces the contents with the required template.

Another might just be the same action, that updates the PR template from branch names which might be more consistent - depends on your teams strategy.

Finally - a Git CLI wizard might offer a url rewrite for the PR URL on a `push`

```
@deitChi ➜ git push
...
remote: 
remote: Create a pull request for 'module/vnet/update-default-dns-servers' on GitHub by visiting:
remote:      https://github.com/ORG/REPO/pull/new/module/vnet/update-default-dns-servers?&template=issue_template.md
remote:
```

**Reply by abstractalgo** [reply]
Those are really decent suggestions, but I still feel this would be easy for them to implement natively in Github and just avoid the excessive workarounds and a still subpar UX.

**Comment by yeseen-emb** [comment]
2025 bump

**Comment by cazlo** [comment]
This is something GitLab supports out of the box https://docs.gitlab.com/user/project/description_templates/

**Comment by FlorianDeconinck** [comment]
One more :+1: for some workflow at NASA :)

**Comment by jaimergp** [comment]
I could see this menu today in `python/peps`:

<img width="740" height="556" alt="image" src="https://github.com/user-attachments/assets/0ebe0eb6-83a0-45db-80ef-cae2c29c4763" />

Is this being rolled out now? 🤔

**Reply by abstractalgo** [reply]
~~They have it mentioned in their [CONTRIBUTING file](https://github.com/python/peps/blob/main/CONTRIBUTING.rst#:~:text=Alternatively%2C%20after%20pushing%20your%20branch%2C%20you%20can%20visit%20https%3A//github.com/python/peps%2C%20and%20follow%20the%20link%20in%20the%20notification%20about%20recent%20changes%20to%20your%20branch%20to%20create%20a%20new%20PR.%20The%20in%2Dbrowser%20interface%20will%20allow%20you%20to%20select%20a%20PR%20template%20for%20your%20new%20PR.) (_"The in-browser interface will allow you to select a PR template for your new PR"_), but I wasn't able to repro in my own repo. Maybe it's under a feature flag of sorts? (Also official Github [docs on PR templates](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) still mention only a single template, so that wasn't updated either)~~

EDIT: see https://github.com/orgs/community/discussions/4620#discussioncomment-15293014

**Reply by uncenter** [reply]
I don't see the menu myself on https://github.com/python/peps/pulls and there aren't any new feature preview toggles on my end.

**Reply by spenserblack** [reply]
> I don't see the menu myself

I'm seeing the dropdown to choose a template. Try this link: https://github.com/python/peps/compare/mal/pep-13-coc-change...main

**Reply by abstractalgo** [reply]
Is anyone able to set up a few .md files in `.github/PULL_REQUEST_TEMPLATE` folder in a sample personal repo and get a similar dropdown for PRs?

**Reply by spenserblack** [reply]
I tried to [here](https://github.com/spenserblack/gengo/tree/bd66e66db8bfbf6b401a96d97613e1c2380c1565/.github/PULL_REQUEST_TEMPLATE), but it didn't seem to work. Feel free to check my work in case I missed something. I wonder if it's only enabled for organizations and/or Team or higher GitHub accounts? Or something else.

**Reply by Andre601** [reply]
I mean, it's pretty obvious that this is most likely a closed alpha phase and the python org is one of the lucky ones that can test it.
So why bother testing? Just wait until there's an announcement about it entering a public beta or being released.

**Reply by uncenter** [reply]
> Is anyone able to set up a few .md files in `.github/PULL_REQUEST_TEMPLATE` folder in a sample personal repo and get a similar dropdown for PRs?

Was not able to in my org repository with https://github.com/catppuccin/userstyles/compare/main...flakey.

**Reply by jaimergp** [reply]
> I mean, it's pretty obvious that this is most likely a closed alpha phase and the python org is one of the lucky ones that can test it.

Confirmed by https://github.com/python/peps/pull/2956. Note it's three years old though, so it doesn't look like this is getting lots of attention.

**Reply by andybeet** [reply]
No it isn't possible. You need to add the name of the template directly to the URL. eg
```
&template=your_template_name.md
```

A drop down is really a needed feature!!

**Comment by canerekare** [comment]
2026 bump

**Comment by tokiya-takai** [comment]
+1 for this, still in 2026. The current `?template=` query parameter approach is effectively unusable in real workflows — no one manually edits URLs to select a template.

While a dropdown UI (like Issue templates) would be a solid improvement, I'd also like to suggest **branch-based auto-selection** as a complementary approach:

Auto-select a PR template based on the **head (source) branch name pattern**. For example:
- `feature/*` → `feature.md` (design intent, screenshots)
- `bugfix/*` → `bugfix.md` (repro steps, root cause)
- `hotfix/*` → `hotfix.md` (impact scope, urgency)
- `release/*` → `release.md` (release checklist, changelog)

Matching against the head branch covers the most common real-world need — different types of work require different checklists, and teams already encode work type in branch naming conventions. Fall back to the default `pull_request_template.md` when no pattern matches.

Ideally, matching against the **base branch** would also be supported for cases like distinguishing PRs targeting `main` vs. a release branch.

Related discussions requesting similar functionality:
- #81319
- #140567
- #165834
